### ISSUE TITLE
To array method in BoolQuery changes _params property to empty StdClass

### ISSUE THREAD

**Issue by JanCizmar** [ruflin/Elastica#1450]
Hello,
I found a serious problem in Query\BoolQuery class.
If method toArray is called, then the _params property is changed to StdClass, but the rest of the code uses the _params propetry like array. So if you want to add some attribute after you call toArray, it dies on error: Cannot use object of type stdClass as array.

```
    public function toArray()
    {
        if (empty($this->_params)) {
            $this->_params = new \stdClass();
        }

        return parent::toArray();
    }
```

Please change this behaviour.

Thanks! :)

**Comment by ruflin** [comment]
Could you share some code snippets from your code? I wonder how / why you access _params directly? Or is this within Elastica?

**Comment by xfoldvar** [comment]
Hey.
We are not accessing _params directly. We are just adding params to a query through method addParam($key, $value) (packages/ruflin/elastica/lib/Elastica/Param.php:137). But this fails when we params are empty.

Exception:
Cannot use object of type stdClass as array
Snippet:

`/** @var $query \Elastica\Query\BoolQuery */`
`$query = $this->elasticSearch->getQuery()->getQuery();`
`$query->addParam("filter", ['bool' => $condition->toElasticQuery()]);`

**Comment by ruflin** [comment]
I tried to figure out what this check for the empty _params is there and the creation of the stdClass. Here is is: https://github.com/ruflin/Elastica/issues/817

@xfoldvar Not sure I can follow your code above. So you get the error when you run `addParam`? Assuming `$query` is a BoolQuery object and toArray is not run yet, I would expect this to work?

**Comment by xfoldvar** [comment]
@ruflin Exactly, the exception is thrown while addParam(key,val) method is called on BoolQuery instance. The property _params is set to stdClass and addParam(key,val) is accessing _params as array.

**Comment by ruflin** [comment]
I still don't understand why _params should be stdClass before `toArray` is called. In the above you seem to use a framework around Elastica. Does the `getQuery()` in there run `toArray()`?

**Comment by enricobono** [comment]
For example here:
https://github.com/ruflin/Elastica/blob/master/src/Query/MatchNone.php
on the constructor we set `_params` to a `stdClass()`

**Comment by ruflin** [comment]
@enricobono Thanks for the specific example, this helps. Lets take the MatchNone as an example. The Elasticsearch query in the end must look as following:

```
{
  "query": {
    "match_none": {}
  }
}
```

If we don't set it to stdClass, it would look as following:

```
{
  "query": {
    "match_none": []
  }
}
```

But this is not valid as match_none has to be an object and not an array. Likely the problem is, that it is not consistent. In some cases it must be an array, in others an object. Do we need to make `addParams` smarter to deal with these differences?

**Comment by enricobono** [comment]
The thing is that when you type:

```
$match = new MatchNone();
$match->setParam('key', 'value');
```

you get an error, because what `->setParam()`  does is:
```
    public function setParam($key, $value)
    {
        $this->_params[$key] = $value;

        return $this;
    }
```
so it tries to add a value to `$this->_params`  which is now a stdClass, so PHP fails with an error:
> Cannot use object of type stdClass as array

**Comment by ruflin** [comment]
@enricobono I understand the general issue but for MatchNone I wonder, what are the params you would be adding to `MatchNone`? Does MatchNone support any?

**Comment by enricobono** [comment]
`MatchNone` is similar to Match, but it negates the params. Something like:

- Match(a, b) -> means get all the records where a == b
- MatchNone(a, b) -> means get all the records where a != b

**Comment by ruflin** [comment]
I've tried to play around with the MatchNone query a bit and the only param I could add was `boost` to get a valid query. Also according to the docs its the opposite of MatchAll: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-all-query.html @enricobono Could you share an example query with MatchNone and params in JSON format that I can try out?

**Comment by MrkMrk00** [comment]
Hello, I would just like to add that this behavior (setting _params to empty \stdClass inside BoolQuery) breaks my use case.
I am introspecting the query from a executed Multi\ResultSet and when I encounter an empty BoolQuery (with no params) I get an Error "Cannot use object of type stdClass as array".

The code (a little different in reality) basically checks, whether the query contains any filters. Something like this:
```{php}
public function isFilterEmpty(): bool
{
        $filterQuery = $this->multiResultSet['main']
            ->getQuery()
            ->getQuery();

        if (!$filterQuery->hasParam('filter') || empty($filterQuery->getParam('filter')) {
            return true;
        }
}
...
```

Calling this function results in:
![image](https://github.com/user-attachments/assets/58392c27-4e47-48a7-bc45-e1eb50a97f9f)

Inside BoolQuery:
```{php}
public function toArray(): array
    {
        if (!$this->_params) {
            $this->_params = new \stdClass();
        }

        return parent::toArray();
    }
```

Since when the query gets "compiled" into a JSON to send to ES, it calls the toArray() method, the query which I get from the ResultSet will have a _params equal to an empty \stdClass.

Not a deal deal breaker, there are other ways to check, just a behavior which I did not expect.

**Comment by ruflin** [comment]
@MrkMrk00 Can you propose how you would "fix/change" the code in Elastica?

**Comment by MrkMrk00** [comment]
I see that there is no "straightforward" fix. My only idea is to not mutate the query but change the type of the resulting array.

Something like this:
```
public function toArray()
{
    $data = parent::toArray();
    if (!$data[$this->_getBaseName()]) {
         $data[$this->_getBaseName()] = new \stdClass();
    }
    
    return $data;
}
```

**Comment by ruflin** [comment]
Do you want to open a PR with the above change? This would also make it visible if it potentially breaks any existing code / tests (I hope not)

**Comment by ruflin** [comment]
Closed in https://github.com/ruflin/Elastica/pull/2241
